### PR TITLE
KAFKA-7798: Expose embedded clientIds

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Add consumer / restoreConsumer / producer(s) / admin client ids via ThreadMetadata; for producerIds, if EOS is turned on add the list of task-producer-ids, otherwise it is a singleton of thread-producer-id.

2. Consolidate the logic of generating clientIds from thread names and client id into StreamThread.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@abbccdda @vvcephei @bbejeck

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 1]
```diff
@@ -1463,31 +1463,31 @@ public void adminClientMetricsVerification() {
 
         final StreamThread.StreamsMetricsThreadImpl streamsMetrics = new StreamThread.StreamsMetricsThreadImpl(metrics, "");
         final StreamThread thread = new StreamThread(
-                mockTime,
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Why we change the style here?

**Review by bbejeck — COMMENTED** [review]
Thanks @guozhangwang  for the patch. Overall looks good to me, but I'm getting some test failures locally on `StreamThreadTest`.

**Review by bbejeck — COMMENTED** [review]
Ignore this comment, I posted twice by accident.

**Comment by guozhangwang** [comment]
@bbejeck That's due to a recent commit that caused conflicts, pushed another commit to resolve it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -736,6 +735,26 @@ private InternalConsumerConfig(final Map<String, Object> props) {
...
+
+    private static String getThreadConsumerClientId(final String threadClientId) {
+        return threadClientId + "-consumer";
+    }
+
+    private static String getThreadRestoreConsumerClientId(final String threadClientId) {
+        return threadClientId + "-restore-consumer";
+    }
+
+    public static String getSharedAdminClientId(final String clientId) {
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:755]: Could we have a comment on explaining what `shared` means in this context?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -1218,17 +1237,43 @@ public final ThreadMetadata threadMetadata() {
...
+            this.state().name(),
+            getThreadConsumerClientId(this.getName()),
+            getThreadRestoreConsumerClientId(this.getName()),
+            producer == null ? Collections.emptySet() : Collections.singleton(getThreadProducerClientId(this.getName())),
+            adminClientId);
+
+        return this;
+    }
+
+    private void updateThreadMetadata(final Map<TaskId, StreamTask> activeTasks,
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1258]: Could we keep the format of `updateThreadMetadata` to avoid unnecessary changes on L1221?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -1218,17 +1237,43 @@ public final ThreadMetadata threadMetadata() {
         return threadMetadata;
     }
 
-    private void updateThreadMetadata(final Map<TaskId, StreamTask> activeTasks, final Map<TaskId, StandbyTask> standbyTasks) {
+    // package-private for testing only
+    StreamThread updateThreadMetadata(final String adminClientId) {
+
+        threadMetadata = new ThreadMetadata(
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1245]: Since this constructor is only called once, could we just reuse the constructor augmented:
```
public ThreadMetadata(final String threadName,
                          final String threadState,
                          final String mainConsumerClientId,
                          final String restoreConsumerClientId,
                          final Set<String> producerClientIds,
                          final String adminClientId,
                          final Set<TaskMetadata> activeTasks,
                          final Set<TaskMetadata> standbyTasks)
```
and set active & standby tasks with empty set? I figure this could reduce code redundancy.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 5]
```diff
@@ -368,20 +368,26 @@ public void setGlobalStateRestoreListener(final StateRestoreListener globalState
...
+            // admin client is shared, so we can actually move it
+            // to result.putAll(adminClient.metrics()).
+            // we did it intentionally just for flexibility.
             result.putAll(thread.adminClientMetrics());
         }
+        // global thread's consumer client
         if (globalStreamThread != null) {
             result.putAll(globalStreamThread.consumerMetrics());
         }
+        // own streams metrics
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: `Own` could sound like a verb. Maybe `self streams metrics` will be better?

**Review by abbccdda — COMMENTED** [review]
LGTM overall. Just left some comments

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: I just did a copy-paste of the code and this is intellij auto formatting. Current style seems right to me.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1258]: I cannot completely follow since L1221 did not change, could you elaborate?

Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1258]: I mean, if we keep parameters in one line:
` private void updateThreadMetadata(final Map<TaskId, StreamTask> activeTasks, final Map<TaskId, StandbyTask> standbyTasks)` then the changes shall be reduced.

**Comment by guozhangwang** [comment]
@mjsax could you give another look on this?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 6]
```diff
@@ -984,10 +982,11 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
             consumerProps.put(entry.getKey(), entry.getValue());
         }
 
-        // add client id with stream client id prefix, and group id
+        // this is a hack to work around StreamsConfig constructor inside StreamPartitionAssignor to avoid casting
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Don't understand this comment? Can you elaborate?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 7]
```diff
@@ -954,8 +954,7 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
      */
     @SuppressWarnings("WeakerAccess")
     @Deprecated
-    public Map<String, Object> getConsumerConfigs(final String groupId,
-                                                  final String clientId) {
+    public Map<String, Object> getConsumerConfigs(final String groupId, final String clientId) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:957]: don't we usually split params over multiple lines?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 8]
```diff
@@ -974,8 +973,7 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
      * @return Map of the consumer configuration.
      */
     @SuppressWarnings("WeakerAccess")
-    public Map<String, Object> getMainConsumerConfigs(final String groupId,
-                                                      final String clientId) {
+    public Map<String, Object> getMainConsumerConfigs(final String groupId, final String clientId) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:976]: ditto

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 9]
```diff
@@ -687,7 +687,8 @@ public static StreamThread create(final InternalTopologyBuilder builder,
             builder,
             threadClientId,
             logContext,
-            assignmentErrorCode);
+            assignmentErrorCode)
+            .updateThreadMetadata(getSharedAdminClientId(clientId));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:691]: Why do we not pass the admin client-id into the constructor?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 10]
```diff
@@ -736,6 +735,26 @@ private InternalConsumerConfig(final Map<String, Object> props) {
...
 
+    private static String getTaskProducerClientId(final String threadClientId, final TaskId taskId) {
+        return threadClientId + "-" + taskId + "-producer";
+    }
+
+    private static String getThreadProducerClientId(final String threadClientId) {
+        return threadClientId + "-producer";
+    }
+
+    private static String getThreadConsumerClientId(final String threadClientId) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: I guess this and the next one could be `getConsumerClientId` and `getRestoreConsumerId` as there is only ever one of each?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1258]: The line is quite long, thus we should reformat it to make the code more readable. I am +1 on this change -- also, I don't think we should optimize for fewer line changes :)

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java [hunk 11]
```diff
@@ -89,7 +89,7 @@ public void shouldThrowExceptionIfBootstrapServersIsNotSet() {
     public void testGetProducerConfigs() {
         final String clientId = "client";
         final Map<String, Object> returnedProps = streamsConfig.getProducerConfigs(clientId);
-        assertEquals(returnedProps.get(ProducerConfig.CLIENT_ID_CONFIG), clientId + "-producer");
+        assertEquals(returnedProps.get(ProducerConfig.CLIENT_ID_CONFIG), clientId);
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java]: shouldn't the arguments be the other way around? i.e., `assertEquals(expected, actual)` ?
better still use `assertThat()` :-P

File: streams/src/main/java/org/apache/kafka/streams/processor/ThreadMetadata.java [hunk 12]
```diff
@@ -73,12 +123,16 @@ public boolean equals(final Object o) {
...
+               mainConsumerClientId.equals(that.mainConsumerClientId) &&
+               restoreConsumerClientId.equals(that.restoreConsumerClientId) &&
+               Objects.equals(producerClientIds, that.producerClientIds) &&
+               adminClientId.equals(that.adminClientId);
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(threadName, threadState, activeTasks, standbyTasks);
+        return Objects.hash(threadName, threadState, activeTasks, standbyTasks, mainConsumerClientId, restoreConsumerClientId, producerClientIds, adminClientId);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/ThreadMetadata.java]: Nit: Can we break this long line into one parameter per line?

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java [hunk 13]
```diff
@@ -98,7 +98,7 @@ public void testGetConsumerConfigs() {
         final String groupId = "example-application";
         final String clientId = "client";
         final Map<String, Object> returnedProps = streamsConfig.getMainConsumerConfigs(groupId, clientId);
-        assertEquals(returnedProps.get(ConsumerConfig.CLIENT_ID_CONFIG), clientId + "-consumer");
+        assertEquals(returnedProps.get(ConsumerConfig.CLIENT_ID_CONFIG), clientId);
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java]: ditto here. I know it was wrong before, but would be good to correct.

**Review by dguy — COMMENTED** [review]
Thanks @guozhangwang left a few minor comments

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 14]
```diff
@@ -101,6 +101,7 @@
     public void before() {
         props = new Properties();
         props.put(StreamsConfig.APPLICATION_ID_CONFIG, "appId");
+        props.put(StreamsConfig.CLIENT_ID_CONFIG, "appId");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: Nit: why reuse `appId` -- if there is a bug, and we set `APPLICATION_ID_CONFIG` by mistake, it would not be detected. Maybe better to set a different id.

**Review by mjsax — COMMENTED** [review]
Couple of nits.

**Review by bbejeck — APPROVED** [review]
LGTM modulo comments from @dguy

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 7 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:957]: It's a subjective question. I personally prefer to have one parameter per line. :)

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:957]: Yeah I personally prefer only doing that for 3+ parameters and they cannot fit in 150 chars which is what a normal laptop screen would fit.

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 6 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Inside `StreamPartitionAssignor` we wrap the passed-in props map into a `InternalStreamsConfig` in order to read its config values without casting. But `InternalStreamsConfig extends StreamsConfig` requires application id as a must-have, so we need to pass in this value to consumer config prop values as well.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:691]: I've thought about passing in the client-id, it is a bit overkill in addition to the thread-client-id just to set the shared admin client inside the constructor. I felt the current way is less intrusive but if you feel strong the other way we can discuss.

File: streams/src/main/java/org/apache/kafka/streams/processor/ThreadMetadata.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/ThreadMetadata.java]: ack.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java (hunk 14 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: good call.

**Comment by guozhangwang** [comment]
Updated per your comments @mjsax @dguy

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 6 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Ack. Makes sense.

However, this is not really clear from the comment. Maybe rephrase to avoid confusion?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 9 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:691]: Because the shared admit client id does never change, it seems weird to have a setter that allow to update is at any time. It seems cleaner to pass it in the constructor and make it immutable this way.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 15]
```diff
@@ -368,20 +368,26 @@ public void setGlobalStateRestoreListener(final StateRestoreListener globalState
...
      */
     public Map<MetricName, ? extends Metric> metrics() {
         final Map<MetricName, Metric> result = new LinkedHashMap<>();
+        // producer and consumer clients are per-thread
         for (final StreamThread thread : threads) {
             result.putAll(thread.producerMetrics());
             result.putAll(thread.consumerMetrics());
+            // admin client is shared, so we can actually move it
+            // to result.putAll(adminClient.metrics()).
+            // we did it intentionally just for flexibility.
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:383]: why not go ahead and do this now?

**Comment by mjsax** [comment]
@guozhangwang I am merging this to cut the release branch. Please address the left over comment in a follow up PR. Thanks.
